File: semana_7_patrones_comportamiento/desarrollo/desarrollo3_command_editor.cpp
/*
 * DESARROLLO 3: Command para editor de texto con guardar y deshacer
 * (EJERCICIO NUEVO - NO es el ejemplo del PDF)
 */

#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Editor;

class Command {
public:
    virtual void ejecutar() = 0;
    virtual void deshacer() = 0;
};

class Editor {
public:
    string contenido;
    vector<Command*> historial;
    
    Editor() : contenido("") {}
    
    void ejecutar(Command* comando) {
        comando->ejecutar();
        historial.push_back(comando);
    }
    
    void deshacer() {
        if (!historial.empty()) {
            Command* comando = historial.back();
            historial.pop_back();
            comando->deshacer();
        }
    }
};

class EscribirCommand : public Command {
private:
    Editor* editor;
    string texto;
    string backup;
    
public:
    EscribirCommand(Editor* e, const string& t) : editor(e), texto(t) {}
    
    void ejecutar() override {
        backup = editor->contenido;
        editor->contenido += texto;
        cout << "Escrito: " << texto << endl;
    }
    
    void deshacer() override {
        editor->contenido = backup;
        cout << "Deshecho" << endl;
    }
};

int main() {
    Editor editor;
    
    EscribirCommand cmd1(&editor, "Hola ");
    EscribirCommand cmd2(&editor, "Mundo");
    
    editor.ejecutar(&cmd1);
    editor.ejecutar(&cmd2);
    cout << "Contenido: " << editor.contenido << endl;
    
    editor.deshacer();
    cout << "Despues de deshacer: " << editor.contenido << endl;
    
    return 0;
}

File: semana_7_patrones_comportamiento/desarrollo/desarrollo3_command_editor.py
"""
DESARROLLO 3: Command para editor de texto con guardar y deshacer
(EJERCICIO NUEVO - NO es el ejemplo del PDF)
"""


class Command:
    def ejecutar(self):
        pass
    
    def deshacer(self):
        pass


class GuardarCommand(Command):
    def __init__(self, editor):
        self.editor = editor
        self.backup = ""
    
    def ejecutar(self):
        self.backup = self.editor.contenido
        print(f"Guardado: {self.editor.contenido}")
    
    def deshacer(self):
        self.editor.contenido = self.backup
        print(f"Restaurado: {self.editor.contenido}")


class EscribirCommand(Command):
    def __init__(self, editor, texto):
        self.editor = editor
        self.texto = texto
        self.backup = ""
    
    def ejecutar(self):
        self.backup = self.editor.contenido
        self.editor.contenido += self.texto
        print(f"Escrito: {self.texto}")
    
    def deshacer(self):
        self.editor.contenido = self.backup
        print(f"Deshecho")


class Editor:
    def __init__(self):
        self.contenido = ""
        self.historial = []
    
    def ejecutar(self, comando):
        comando.ejecutar()
        self.historial.append(comando)
    
    def deshacer(self):
        if self.historial:
            comando = self.historial.pop()
            comando.deshacer()


editor = Editor()
editor.ejecutar(EscribirCommand(editor, "Hola "))
editor.ejecutar(EscribirCommand(editor, "Mundo"))
print("Contenido:", editor.contenido)
editor.deshacer()
print("Despues de deshacer:", editor.contenido)
